core/src/com/fourinachamber/fortyfive/screen/general/AdvancedTextWidget.kt

```diff
@@ -7,78 +7,126 @@ import com.badlogic.gdx.graphics.g2d.GlyphLayout
 
-    private var nodesOfCurrentText: List<YogaNode> = listOf()
+    override var fixedZIndex: Int = -1
 
-    open var advancedText: AdvancedText = advancedText
+    override var isHoveredOver: Boolean = false
+    override var isClicked: Boolean = false
+    override var styleManager: StyleManager? = null
+
+    open var advancedText: AdvancedText = AdvancedText.EMPTY
         set(value) {
             field = value
             clearText()
-            initNodes(value)
+            initText(value)
         }
 
+    private var initialisedStyleInstruction: Boolean = false
+
     init {
-        initNodes(advancedText)
+        @Suppress("LeakingThis")
+        bindHoverStateListeners(this)
+        @Suppress("LeakingThis")
+        initText(advancedText)
+    }
+
+    fun setRawText(text: String) {
+        advancedText = AdvancedTextParser(text, screen, defaults).parse()
     }
 
-    private fun initNodes(value: AdvancedText) {
-        nodesOfCurrentText = value.parts.map { part ->
-            val node = add(part.actor)
-            if (part is StyledActor && part is Actor) { // TODO: .actor is stupid now
-                val oldManager = part.styleManager
-                val newManager = StyleManager(part, node)
-                part.styleManager = newManager
-                part.initStyles(screen)
-                newManager.addInstruction( // TODO: this is ugly
-                    "marginLeft",
-                    StyleInstruction(
-                        YogaValue(0.1f, YogaUnit.POINT),
-                        1,
-                        StyleCondition.Always,
-                        YogaValue::class
-                    ),
-                    YogaValue::class
-                )
-                if (oldManager == null) {
-                    screen.addStyleManager(newManager)
-                } else {
-                    screen.swapStyleManager(oldManager, newManager)
+    override fun layout() {
+        super.layout()
+
+        children
+            .filterIsInstance<Layout>()
+            .onEach(Layout::validate)
+            .forEach { child ->
+                child as Actor
+                child.width = child.prefWidth
+                child.height = child.prefHeight
+            }
+
+        var curX = 0f
+        val lines = advancedText
+            .parts
+            .splitAt { part ->
+                val child = part.actor
+                val shouldSplit = curX + child.width > width
+                if (shouldSplit) {
+                    curX = 0f
+                } else if (part.breakLine) {
+                    curX = width + 1f
                 }
+                curX += child.width
+                shouldSplit
             }
-            node
-        }
+            .map { line -> line.map { it.actor } }
+
+        var curY = 0f
+        curX = 0f
+        lines
+            .reversed()
+            .zip { line -> line.maxOf { it.height } }
+            .forEach { (line, height) ->
```

Comment:
the command:
"line.maxOf(it.height)"
can be extracted into a variable to not calculate it more than once

Reply:
line is the lambda paramenter, which is different each iteration